crates/bevy_render/src/render_phase/mod.rs

```diff
@@ -51,18 +52,254 @@ use std::{ops::Range, slice::SliceIndex};
 /// the rendered texture of the previous phase (e.g. for screen-space reflections).
 /// All [`PhaseItem`]s are then rendered using a single [`TrackedRenderPass`].
 /// The render pass might be reused for multiple phases to reduce GPU overhead.
+///
+/// This flavor of render phase is used for phases in which the ordering is less
+/// critical: for example, `Opaque3d`. It's generally faster than the
+/// alternative [`SortedRenderPhase`].
 #[derive(Component)]
-pub struct RenderPhase<I: PhaseItem> {
+pub struct BinnedRenderPhase<BPI>
+where
+    BPI: BinnedPhaseItem,
+{
+    /// A list of `BinKey`s for batchable items.
+    ///
+    /// These are accumulated in `queue_material_meshes` and then sorted in
+    /// `batch_and_prepare_binned_render_phase`.
+    pub batchable_keys: Vec<BPI::BinKey>,
+
+    /// The batchable bins themselves.
+    ///
+    /// Each bin corresponds to a single batch. For unbatchable entities, prefer
+    /// `unbatchable_values` instead.
+    pub(crate) batchable_values: HashMap<BPI::BinKey, Vec<Entity>>,
```

Comment:
it may allocate many small memory in some cases, Could smallvec used here?

Reply:
I'd prefer merging it as is then experiment with SmallVec or something like that in a future PR.

Reply:
I tried `SmallVec` and it didn't improve performance.